### ISSUE TITLE
Streaming always stops after a few hours and can't be cancelled

### ISSUE THREAD

**Issue by stmax82** [JoeMayo/LinqToTwitter#122]
I'm trying to work with the streaming API (type "sample" and "filter").

The stream works for several hours and then just stops receiving messages. It doesn't throw exceptions and the "await" doesn't return either.

It happens with the following piece of code:

```csharp
using System;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;

using LinqToTwitter;

namespace TwitterStreamRecorder {
    public class Program {
        public static async Task<int> Main(string[] args) {
            var auth = new SingleUserAuthorizer {
                CredentialStore = new SingleUserInMemoryCredentialStore {
                    ConsumerKey = "...",
                    ConsumerSecret = "...",
                    AccessToken = "...",
                    AccessTokenSecret = "..."
                }
            };

            while (true) {
                try {
                    using (var ctx = new TwitterContext(auth)) {
                        await ctx
                        .Streaming
                        .Where(x => x.Type == StreamingType.Sample)
                        .StartAsync(strm => {
                            Console.WriteLine(strm.Content.Substring(0, 100) + "...");
                            return Task.CompletedTask;
                        });
                    }
                }
                catch (Exception ex) {
                    Console.WriteLine("Exception: " + ex.Message);
                }

                Thread.Sleep(TimeSpan.FromSeconds(60));
            }
        }
    }
}
```

I tried to work around the problem by providing a cancellation token to the "WithCancellation" method. The cancellation token gets cancelled when no messages were received for 15 seconds. See code below:

```csharp
using System;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;

using LinqToTwitter;

namespace TwitterStreamRecorder {
    public class Program {
        public static async Task<int> Main(string[] args) {
            var auth = new SingleUserAuthorizer {
                CredentialStore = new SingleUserInMemoryCredentialStore {
                    ConsumerKey = "...",
                    ConsumerSecret = "...",
                    AccessToken = "...",
                    AccessTokenSecret = "..."
                }
            };

            while (true) {
                try {
                    using (var ctx = new TwitterContext(auth))
                    using (var cancel = new CancellationTokenSource())
                    using (var watchdog = new Watchdog(cancel, TimeSpan.FromSeconds(15))) {
                        await ctx
                        .Streaming
                        .WithCancellation(cancel.Token)
                        .Where(x => x.Type == StreamingType.Sample)
                        .StartAsync(strm => {
                            Console.WriteLine(strm.Content.Substring(0, 100) + "...");
                            watchdog.Heartbeat();
                            return Task.CompletedTask;
                        });
                    }
                }
                catch (OperationCanceledException) {
                    Console.WriteLine("Watchdog activated. Retrying...");
                }
                catch (Exception ex) { 
                    Console.WriteLine("Exception: " + ex.Message);
                }

                Thread.Sleep(TimeSpan.FromSeconds(60));
            }
        }
    }

    public class Watchdog : IDisposable {
        private readonly System.Timers.Timer timer;

        private volatile int count;

        public Watchdog(CancellationTokenSource cancel, TimeSpan timeout) {
            timer = new System.Timers.Timer {
                AutoReset = true,
                Interval = timeout.TotalMilliseconds
            };

            int prevCount = count;

            timer.Elapsed += (s, e) => {
                if (prevCount != count)
                    prevCount = count;
                else {
                    Console.WriteLine($"Watchdog timeout. Cancelling...! ({prevCount}, {count})");
                    cancel.Cancel();
                }
            };

            timer.Start();
        }

        public void Heartbeat() => count++;
        public void Dispose() => timer.Dispose();
    }
}
```

This also receives messages for several hours, then it stops and prints "Watchdog timeout. Cancelling...!" after 15 seconds. But **even though the cancellation token has been cancelled, the await never returns** (it never prints "Watchdog activated. Retrying...").

Tested with linqtotwitter 4.2.1 on Win 10, VS 15.6.4, C# 7.2.

**Comment by Badabum** [comment]
Same thing. @stmax82 have you been able to resolve this? @JoeMayo Do you have any suggestions?

**Comment by stmax82** [comment]
@Badabum no we didn't find a way to resolve this. We switched to Python and the python-twitter library from https://github.com/bear/python-twitter. It works incredibly stable 24x7 since more than a year now!

**Comment by Badabum** [comment]
@stmax82 thanks for the answer!

**Comment by JoeMayo** [comment]
I've been running a sample stream over the last 24 hours to test and see if I can reproduce. It's still running without problems.

Doing further research, I looked at the code history and noticed that around the October 2019 time frame, I added a `ConfigureAwait(false)` on one of the async methods in the streaming call path. This was on LINQ to Twitter v5.0. Without that `ConfigureAwait`, there was potential for a threading issue.

If anyone sees any more problems, I'll take a look.
